riscv/devices.h

```diff
@@ -41,21 +41,50 @@ class mem_t : public abstract_device_t {
  public:
   mem_t(reg_t size);
   mem_t(const mem_t& that) = delete;
-  ~mem_t();
+  virtual ~mem_t() = default;
 
   bool load(reg_t addr, size_t len, uint8_t* bytes) { return load_store(addr, len, bytes, false); }
   bool store(reg_t addr, size_t len, const uint8_t* bytes) { return load_store(addr, len, const_cast<uint8_t*>(bytes), true); }
-  char* contents(reg_t addr);
   reg_t size() { return sz; }
-  void dump(std::ostream& o);
+  virtual char* contents(reg_t addr) = 0;
+  virtual void dump(std::ostream& o) = 0;
+
+ protected:
+  reg_t sz;
 
  private:
-  bool load_store(reg_t addr, size_t len, uint8_t* bytes, bool store);
+  virtual bool load_store(reg_t addr, size_t len, uint8_t* bytes, bool store) = 0;
+};
+
+class sparse_mem_t : public mem_t {
+ public:
+  sparse_mem_t(reg_t size);
+  ~sparse_mem_t();
+
+  char* contents(reg_t addr) override;
+  void dump(std::ostream& o) override;
```

Comment:
You added `override` but not `virtual`. Please add `virtual` to be consistent with other code in Spike

Reply:
Thank you, addressed in [`69e70ad` (#1342)](https://github.com/riscv-software-src/riscv-isa-sim/pull/1342/commits/69e70ad0bdbbe68d9f3fc2868f46174da7e49478)